Support single channel

zagrodzki opened the pull request:
With the custom firmware on the HT6022BE, it's possible to disable one channel, allocating more USB bandwidth for the first channel and achieving higher effective sample rate. Tests show successful 24msps with isochronous transfers.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zagrodzki/goscope/36)
<!-- Reviewable:end -->

zagrodzki committed:
When a capture error is reported, abort the program instead of silently
ignoring.

zagrodzki committed:
Support both single and dual channel config.

zagrodzki committed:
Support single or dual channel configuration.

zagrodzki committed:
Quick&dirty flags for controlling basic capture params.

zagrodzki committed:
Support single and dual channel configuration.

zagrodzki committed:
Remove unnecessary printing.

zagrodzki committed:
Replace fmt.Errorf with errors.Errorf for consistency...

zagrodzki committed:
Merge remote-tracking branch 'github/master' into support_single_channel

zagrodzki commented:
@dtkaczyk

zagrodzki committed:
Number of channels needs to be initialized to 2 in case the device
runs the original firmware - we don't then issue the setNumChannels
request.
Check the errors returned by set* methods during init.

zagrodzki committed:
Wrong argument type for printf, voltrange is an int

dtkaczyk commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[usb/hantek6022be/capture.go, line 39 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYmHMlgkk72dlvsraX:-KnYmHMmuvDMSPr4xCzt:b-nqd954) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L39)):*
> ```Go
> 	if recTB := h.rec.TimeBase(); tb > 8*recTB {
> 		tb = 8 * recTB
> 	}
> ```

Since tb is used only if !h.iso, these 4 lines could be moved inside "case false" below.

---

*[usb/hantek6022be/capture.go, line 43 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYmYwwfu7a5YBSH9Js:-KnYmYwwfu7a5YBSH9Jt:bg8trs7) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L43)):*
> ```Go
> 	switch h.iso {
> 	case true:
> 		readLen = 3072 * 255
> ```

It is unclear why readLen is constant here and where do these numbers come from. Maybe add a comment?

---

*[usb/hantek6022be/capture.go, line 75 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnY8RrfD1PWvcUhfgbo:-KnY8Rrgg-mjJacSGztm:bkbm796) ([raw file](https://github.com/zagrodzki/goscope/blob/7b18196f42019fa3db70120be0371694d6c8c3c8/usb/hantek6022be/capture.go#L75)):*
> ```Go
> 		return errors.Wrap(err, "Read")
> 	}
> 	if num > len(sampleBuf) {
> ```

Why this check disappeared?

---

*[usb/hantek6022be/capture.go, line 78 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnY85uXL3dYqFS95vb0:-KnY85uYSnYlGoAcm1rT:b-ezbim9) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L78)):*
> ```Go
> 		return errors.Wrap(err, "Read")
> 	}
> 	if num%int(h.numChan) != 0 {
> ```

Why casting h.numChan into int? Scope.numChan is already defined as an int.

---

*[usb/hantek6022be/capture.go, line 79 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYBdzJRCOcpV5YSlPg:-KnYBdzKdqEhjltHT4ck:bozfp1z) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L79)):*
> ```Go
> 	}
> 	if num%int(h.numChan) != 0 {
> 		return errors.Errorf("Read returned %d bytes of data, expected an even number for 2 channels", num)
> ```

It's a bit confusing that the code is super general and prepared for any number of channels (not only 1 or 2), but the error message covers only one specific case of 2 channels. I get that in practice this condition will be true only if there are 2 channels and odd number of bytes, but maybe the message should be general as the rest of the code?

---

*[usb/hantek6022be/init.go, line 79 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYaZp_z1Pzn_7nZiuI:-KnYaZpafjJFbjH4hsnZ:b-xnqcpo) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/init.go#L79)):*
> ```Go
> 			return nil, fmt.Errorf("setNumChan(%d): %v", numChan, err)
> 		}
> 	}
> ```

I don't really understand what does number of channels have to do with type of transfer (iso/bulk). If bulk is used, numChan is not set - why?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/36)*
<!-- Sent from Reviewable.io -->

zagrodzki committed:
Set better buffer size for the ISO case.
Make comment more generic.

zagrodzki commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[usb/hantek6022be/capture.go, line 39 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYmHMlgkk72dlvsraX:-KnZPagbYOwgYxD3-qOV:b-3cve5k) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L39)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

Since tb is used only if !h.iso, these 4 lines could be moved inside "case false" below.
</blockquote></details>

Actually, it makes sense to use tb for h.iso too, since we can still use very low sample rates (100ks/s) which would result in 3072*255 returning ~7s of data. Which doesn't make sense. Changed.

---

*[usb/hantek6022be/capture.go, line 43 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYmYwwfu7a5YBSH9Js:-KnZPs1nsj6VcxubBl4W:b-h3oxa8) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L43)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

It is unclear why readLen is constant here and where do these numbers come from. Maybe add a comment?
</blockquote></details>

Changed, see above. Added a comment about "3072" too.

---

*[usb/hantek6022be/capture.go, line 78 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnY85uXL3dYqFS95vb0:-KnZQ7ZZ0RSKirHYRaJk:b-896fix) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L78)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

Why casting h.numChan into int? Scope.numChan is already defined as an int.
</blockquote></details>

Done.

---

*[usb/hantek6022be/capture.go, line 79 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYBdzJRCOcpV5YSlPg:-KnZQHgDVcGdle07IiCB:b-896fix) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L79)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

It's a bit confusing that the code is super general and prepared for any number of channels (not only 1 or 2), but the error message covers only one specific case of 2 channels. I get that in practice this condition will be true only if there are 2 channels and odd number of bytes, but maybe the message should be general as the rest of the code?
</blockquote></details>

Done.

---

*[usb/hantek6022be/init.go, line 79 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYaZp_z1Pzn_7nZiuI:-KnZQKHuVC93N1Nfuaip:boy2tfi) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/init.go#L79)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

I don't really understand what does number of channels have to do with type of transfer (iso/bulk). If bulk is used, numChan is not set - why?
</blockquote></details>

There are two different firmwares for the HT6022BE, which effectively make it behave like two slightly different devices.
With the original vendor firmware, it always sends data for two channels, no way to ask it for just single channel data. It only supports bulk data transfers.
With the opensource firmware, it has an option to select number of channels. It also supports isochronous data transfers, which are more reliable than bulk transfers - there's guaranteed bandwidth and soft-guaranteed latency, where bulk transfers might be delayed if there's other activity on the bus. This is why the flag is called "iso", but it's a bad name and I'll change it soon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/36)*
<!-- Sent from Reviewable.io -->

zagrodzki commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[usb/hantek6022be/capture.go, line 75 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnY8RrfD1PWvcUhfgbo:-Kn__-nRSCI-hoB63JGL:b-zfv941) ([raw file](https://github.com/zagrodzki/goscope/blob/7b18196f42019fa3db70120be0371694d6c8c3c8/usb/hantek6022be/capture.go#L75)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

Why this check disappeared?
</blockquote></details>

which check?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/36)*
<!-- Sent from Reviewable.io -->

zagrodzki commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[usb/hantek6022be/capture.go, line 75 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnY8RrfD1PWvcUhfgbo:-Kn__edPHPQLOj8v3KJf:bdsifs8) ([raw file](https://github.com/zagrodzki/goscope/blob/7b18196f42019fa3db70120be0371694d6c8c3c8/usb/hantek6022be/capture.go#L75)):*
<details><summary><i>Previously, zagrodzki wrote…</i></summary><blockquote>

which check?
</blockquote></details>

ah, this check. It was added some time ago only as I was trying to debug an issue with sample buffer being not consistent with USB read buffer. I think the original issue is now resolved and we never hit that condition anymore - the buffer length is passed to libusb for a read operation, so libusb transaction should never return a number larger than the buffer size. At least that's what I remember.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/36)*
<!-- Sent from Reviewable.io -->

zagrodzki commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[usb/hantek6022be/capture.go, line 39 at r3](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-KnYmHMlgkk72dlvsraX:-Kn_grzNqWpabnslBjjH:b-896fix) ([raw file](https://github.com/zagrodzki/goscope/blob/7ce0af9bc3a6fcbcb70c2e121a75b5246d985d2f/usb/hantek6022be/capture.go#L39)):*
<details><summary><i>Previously, zagrodzki wrote…</i></summary><blockquote>

Actually, it makes sense to use tb for h.iso too, since we can still use very low sample rates (100ks/s) which would result in 3072*255 returning ~7s of data. Which doesn't make sense. Changed.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/36)*
<!-- Sent from Reviewable.io -->

dtkaczyk commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 7 files at r1, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/36#-:-Kn_lEkPczmVVwHwQl-B:bnfp4nl)*
<!-- Sent from Reviewable.io -->
